### PR TITLE
Profiling framework: core performance-logging implementation

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Implement [#181038002: Profiling framework: core performance-logging implementation](https://www.pivotaltracker.com/n/projects/2539304/stories/181038002).

### Important Notes

- ~An unanticipated complexity: for a proc macro to create a wrapper, it needs to generate a call to the wrapped function, based on the input signature. Doing this for any possible signature would be complicated, and I don't see any other crate that solves the problem already. For the initial implementation, I have included support for basic `fn` and `async fn` function and method calls. What I have not implemented is support for all the possible destructuring-binds that can be performed in function signatures. Destructuring in signatures isn't common in the Enso codebase. If someone does apply `#[profile]` to a function with such a signature, it will fail at compile time with a note that that isn't supported; rewriting the function to work without destructuring in the signature is trivial.~ [Edit: the wrapper doesn't have to have the same bindings as the callee, just the same number and type of arguments; we can bind everything to arbitrary idents]
- For performance, I have used `unsafe` in two different areas: for NonZeroU64 in Timestamp (sound as long as Timestamp doesn't wrap, and it won't for 53 million years), and to support shared mutable access to `profiler::MEASUREMENTS`. The latter is to avoid the overhead of a RefCell check for every logged measurement; soundness of the construct is ensured by a simple rule (documented in a source note).
- ~Due to reliance on JS `performance.now()`, the implemented tests are WASM tests. `run`'s test support should identify them automatically; to test the crate by hand, you can use `wasm-pack test --node`~ [Edit: mock implementation]
- I've used an environment variable to set profiling level, e.g. `ENSO_MAX_PROFILING_LEVEL=task`. This avoids needing to propagate cargo feature flags between crates.
- ~CLion doesn't seem to be able to display the expansion of the proc macro, but cargo-expand works if you'd like to see the generated code.~

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

[ci no changelog needed]

### CODE REVIEW

File: Cargo.toml
Comment (wdanilo) [inline Cargo.toml]: We are normally putting macros and implementation in a common folder. Here that should be `lib/rust/profiler/macros` and `lib/rust/profiler/impl`. This way, such layout is more future-proof. For example, if we would like to extract that as a separate library to its dedicated repository in the future, the `lib/rust/profiler` would become the repository `src` folder. We have the same layout already in `lib/rust/parser/flexer-testing`, although the naming there is not the best, and I think that we should stay with `/macros` and `/impl` naming instead (I'm very open for other name proposals!).

File: lib/rust/profiler-macros/Cargo.toml
Comment (wdanilo) [inline lib/rust/profiler-macros/Cargo.toml]: no newline

File: lib/rust/profiler-macros/Cargo.toml
Comment (wdanilo) [inline lib/rust/profiler-macros/Cargo.toml]: 1. Please add the `authors` section just as in other Cargo.toml files.
2. Our convention here is to prefix the names of the packages with `enso-`, so this package would be named `enso-profiler-macros`. We are using this convention with other libraries too, like `prelude`, `optics`, `logger`, etc. The idea is that this way we have a way smaller chance of crashing with a name of a library registered on crates.io and until we publish these libraries as separate crates, we keep their names prefixed.

File: lib/rust/profiler-macros/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler-macros/src/lib.rs]: Please make sure that all comments are valid English sentences starting with a capital letter and describing what the thing does. I understand what you mean here, but describing that a little bit more could help readers to easily read comments.

File: lib/rust/profiler-macros/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler-macros/src/lib.rs]: Such macros should have a much more detailed documentation, especially in the situation when IntelliJ / CLion are unable to expand them in-place. This should have a description of what it does, what are its usages, what are its limitations (like no support for arg pattern matches), and also, it should show the expanded function for a particular  example function.

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/src/lib.rs]: This macro is not proc-macro. Is the expansion understood by IntellIJ / CLion – are the created structures hinted to the user? For example, if I write `profiler::` am I hinted with different possible profiler hierarchy definitions? It is important to hint the user of this API as much as possible. If these are not hinted to the user, I would vote for moving the generation of the new structures to proc-macros and leaving only the `#[profile]` macro as a non-proc-macro definition.

Also, as we've been speaking during designing of this task, in case the user is not hinted with anything when using IntelliJ, we should just use proc-macros even if they are uglier. It applies to the `#[profile]` vs `profile!(...)` syntax as well. If the `#[profile]` syntax prevents the user of this API from any hints, we should use proc-macros instead.

File: lib/rust/profiler-macros/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler-macros/src/lib.rs]: Please use the correct Rust doc syntax when referring to code structures, like ``` [`Measurement`] ```.

File: lib/rust/profiler-macros/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler-macros/src/lib.rs]: No module doc. Probably the linter of this crate is not configured a linter of other crates – please copy the configuration from other `lib.rs`. Unfortunately, it is impossible to create a single linter configuration for all the crates :(

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/src/lib.rs:148]: No module doc – this doc is super important here. This is the root of the library and this doc should be something a new team member or a new user of this library will read to understand how the whole library works. It should be a **complete** introduction to the profiler framework.

This doc should be a lengthy one – it should provide the user with as much information as possible. In particular, it should contain info about motivation, usage patterns, performance, disabling and enabling profilers, the guidance (when to use which profiler), etc. In fact, the doc here should be a shorter version of the design document and it should contain sections and should be nicely layouted.

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/src/lib.rs]: Amazing explanation and cool design, thanks for providing the links!

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/src/lib.rs]: in both equations you are using the a little bit magical `+ 1` and `- 1`. Could you extract the `1` constnt to a well named constant that describes its semantics, please?

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/src/lib.rs:285]: missing newline

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/src/lib.rs]: There is one function referring this note below the note and one above the note. Move the note below both of them for just a tiny better clarity when reading the code, please :)

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/src/lib.rs:350]: This is a cool structure! Maybe it should be extracted to `enso/lib/rust/data-structures`?

**Review by wdanilo — CHANGES_REQUESTED** [review]
Really amazing implementation!

1. The code quality is really nice. I'm really surprised how nicely it follows all of our code guidelines, especially, as this is your first bigger PR. 
2. The description of this PR is amazing too!
3. Although I normally hate using `unsafe` bits, your usage is very valid, nicely described and I'm glad you came up with such a nice solution.
4. There are no `bench` tests, and I think we should add them. I'd do them as follow – I would create a function `performance_now()` which uses JS if compiled to WASM and uses whatever when compiled to native platform (it can even return 0). Then, we should create benchmarks running natively and they should compare the performance of our profiling framework with for example pushing numbers to a `Vec` (so we will see how big the slowdown is in comparison to just `Vec` operations). We do not want to use WASM tests / benchmarks, as they are killing our CI, and we are moving more and more code from WASM tests to native tests.
5. Would it be possible to use the trick described in point 3 to move the tests from WASM tests to native tests? Our long-term goal is to have 0 WASM tests and make the CI running really fast.
6. I left few comments here and there about lack of documentation – this is a pretty important issue with this PR to me. It's easy to be solved, however, it's insanely important that a new library provides the developer with a complete documentation and guidance of how to use the library. It should contain all the details, including such things as the usage of env var `ENSO_MAX_PROFILING_LEVEL` – I would never know to use it based on the current code / docs!
7. Another important topic that I left comments about are the macros – I am afraid that some of the macros prevent IntelliJ from providing the user of this library hints about available structures or methods. See the comments to learn more.
8. What logger is used when building this pkg if use does not set `ENSO_MAX_PROFILING_LEVEL` explicitly?

Overall, this is absolutely amazing code! ❤️

File: Cargo.toml
Comment (farmaazon) [reply Cargo.toml]: I'd prefer to _not_ make "impl" directory; instead, let the implementation be placed in lib/rust/profiler and macros be a "sub-crate" in lib/rust/profiler/macros. This is more intuitive when filling up the dependencies (I constantly forget putting "impl" in paths).

File: lib/rust/profiler/src/lib.rs
Comment (kazcw) [reply lib/rust/profiler/src/lib.rs]: 1. This *is* a proc macro--I think you mean it is not a declarative macro. Declarative macros are the simpler `macro_rules!` macros; proc macros are the kind that explicitly operate on a `TokenStream`, that JetBrains only recently supports.
2. It is [working for me](https://i.imgur.com/cTUKSd3.png), in CLion 2021.3.1 (released Dec 28). Are you using the latest release of IntelliJ?

**Comment by kazcw** [comment]
Added benchmarks.

Performance impact: Except at low numbers of measurements, run time is dominated by growing the vector. I'm measuring about 1.6ns per logged measurement, when accumulating 10k measurements. I think the cost of the unavoidable `performance.now()` will be on the order of 1μs, in which case our overhead is within 0.1% of an optimal implementation.

Performance variability impact: There's no easy way to measure this, so I'm speaking theoretically here. The only operation expected to have a significantly variable cost is the `Vec::push` to grow the `MEASUREMENTS` log; it sometimes needs to reallocate. However even at its most expensive, it should be on the order of a 1μs (for reasonable numbers of measurements); so the variance introduced by this framework shouldn't disturb even very small measurements (I expect <1% added variability for a 1ms measurement).

**Comment by kazcw** [comment]
I need to change the structure of the wrapped code generated by `#[profile]` so that it is compatible with any context (e.g. the current implementation won't work in a trait impl). Will do that tomorrow, the fix is straightforward.

File: lib/rust/profiler/src/lib.rs
Comment (wdanilo) [reply lib/rust/profiler/src/lib.rs]: 1. Ugh, I'm always mixing the names here. OF course, I meant `macro-rules` here instead.
2. I was not testing it TBH. I know it was not working a few months ago, so I preferred to ask. If it works now, then great! (I hope you did not read my message as "it does not work, lets fix it!" – it was a question about IF it works with explanation why its important to me) :)

**Comment by wdanilo** [comment]
> Added benchmarks.
> 
> Performance impact: Except at low numbers of measurements, run time is dominated by growing the vector. I'm measuring about 1.6ns per logged measurement, when accumulating 10k measurements. I think the cost of the unavoidable `performance.now()` will be on the order of 1μs, in which case our overhead is within 0.1% of an optimal implementation.
> 
> Performance variability impact: There's no easy way to measure this, so I'm speaking theoretically here. The only operation expected to have a significantly variable cost is the `Vec::push` to grow the `MEASUREMENTS` log; it sometimes needs to reallocate. However even at its most expensive, it should be on the order of a 1μs (for reasonable numbers of measurements); so the variance introduced by this framework shouldn't disturb even very small measurements (I expect <1% added variability for a 1ms measurement).

Perfect! Two additional questions:
1. Could you write these notes next to the benchmark suite as a comment? It would be amazing to read your findings when reading the code.
2. I've got an idea but I'm not sure yet what to think about it – what about creating a test that runs both `Vec::push` and our framework and if our framework runs longer than `x * time_of_vc_push` it fails? This way we will catch the situation when someone creates some changes to the code which break the perf. I believe that it would be possible to find an `x` which is big enough to not be affected by the time variance and small enough to catch bad modifications of this codebase in the future. What do you think about it? Do you think it's doable / worth it?

**Review by wdanilo — APPROVED** [review]
This code is already in a mergeable shape, so I'm accepting it. The new documentation in the module and overall code quality are really high. Amazing job with it, Kaz!

File: lib/rust/profiler/src/lib.rs
Comment (kazcw) [reply lib/rust/profiler/src/lib.rs]: (Ignore this, I did more testing, will update below in a top-level comment)

---

So, **`define_hierarchy!` is fully working** for me: CLion seems to know the expansion of that proc macro; it suggests the generated types, has documentation for them, etc.

However I just noticed a problem with `#[profile]`, and using the `profile!` syntax doesn't help: *within* the decorated function, hints are working as usual; but the IDE doesn't seem to know about the *output*, i.e. **adding profiling to a function causes the IDE no longer to know about the function's definition**. This is [supposed to be working](https://github.com/intellij-rust/intellij-rust/issues/6908), but it looks like there are still gaps in their implementation.

# What does JetBrains support?

Doing some testing, it looks like the IDE understands *most* items generated by a proc macro, including everything `define_hierarchy!` needs:
- structs
- macros
- functions defined in impl blocks

What isn't working is proc macros that expand to *a single function*, rather than an item that can contain functions.

# Alternative APIs that avoid the bug

## Use `profile!(fn ..)` syntax; implement with a declarative macro

Pros:
- It would work with current tooling. (The first step would be to do a mock implementation to verify this).

Cons:
- The syntax is not as nice.

Complexity: I think a declarative implementation that supports everything we need to support would be tricky. I could probably implement it in 3-4 days.

## Apply `#[profile]` to the outer objects

JetBrains seems to be okay with macros generating impl blocks, so we could apply `#[profile]` at a higher scope than individual `fn`s.

Example:
```
struct Foo;

#[profile]
impl Foo {
    fn profiled_fn1(&self, [profiler: profiler::Task]) { .. }
    fn profiled_fn2(x: u32, [profiler: profiler::Objective]) { .. }
    fn not_profiled(self, x: u32) { .. }
}
```

(In this syntax, profiler arguments are marked with brackets, to make it more obvious where the macro is changing signatures).

Pros:
- It would work with current tooling. (I have already tested this: the IDE knows about functions emitted inside struct impls, trait impls, and modules).
- Unobtrusive syntax
- When JetBrains fixes the bug, we could also enable annotating at `fn` granularity

Cons:
- Moving the attribute farther from the functions it affects makes it less obvious what's going on with the `profiler` arguments.

Complexity: This would be simpler than the declarative implementation; I think I could implement and document it 2 days.

# Proceeding

For `#[profile]`, we have design change and implementation work needed. The lower-level `define_hierarchy!` is stable; I don't anticipate needing any changes to accommodate whatever we do with `#[profile]`. Should I cut the `#[profile]` code out of this PR so we can get the low-level code merged, and tackle the high-level interface separately? (If we go with the declarative `profile!` API, that will require a rewritten implementation anyway)

**Comment by kazcw** [comment]
# JetBrains IDE status

I've just tested these macros with different versions of JetBrains IDEs. ~~(I was surprised to find that IntelliJ`s current macro bugs are completely non-overlapping with CLion's.)~~

|  | `#[profile] fn` |  `define_hierarchy!` |
| --- | --- | --- |
| experimental features disabled |  :white_check_mark:  | :x: |
| experimental features enabled |  :x:  | :heavy_check_mark: |
| occasionally, experimental features enabled |  :grey_question:  | :heavy_check_mark: |

:x: = IDE doesn't know about items generated by the macro
:white_check_mark: = IDE ignores macro; usable but not perfect
:heavy_check_mark: = working

# ~~Here's what we can do~~

~~To avoid missing docs in anyone's IDE:~~
~~1. Everyone will need to enable experimental features~~
~~2. for IntelliJ users: **expand `define_hierarchy!` (e.g. with cargo-expand), and check the results into source**. This macro expansion step must be repeated any time we make changes to the `define_hierarchy!` implementation, but I don't anticipate needing to do that often. Once IntelliJ can handle the macro, we can remove the generated code from git and run the macro directly.~~

~~@wdanilo how does this sound for IDE support?~~

**Comment by mwu-tow** [comment]
> I've just tested these macros with different versions of JetBrains IDEs. (I was surprised to find that IntelliJ`s current macro bugs are completely non-overlapping with CLion's.)

The IDE should not affect this. Could you see if they are running the same version of the Rust plugin and the configuration (including experimental features) is the same?
If so, and you see any difference, this is something I'd consider of being worth of reporting to the Rust plugin developers. 

Well, actually any failures with latest plugin version with experimental macro expansion enabled should be reported IMHO, they are currently patching this feature quite actively.

**Comment by kazcw** [comment]
I found the root of the problem: using `proc_macro::Span` methods to obtain file/line info (which requires the `proc_macro_span` crate feature) causes the IDE's macro expansion to fail with `not implemented`. We use those methods to build the measurement's label, which doesn't affect the information the IDE needs from macro-expansion (signatures and docs). So, I'll add a feature to the crate that enables/disables attaching file/line info to measurements; by default it is disabled (so the IDE can build the crate), but the `run` build commands will enable the feature.

(The variation in results I thought was caused by editor version seems to have been caused by something else.)

File: lib/rust/profiler/src/lib.rs
Comment (kazcw) [reply lib/rust/profiler/src/lib.rs:350]: I'll do that. Another reusable component of this: `wrap_body` could easily be used to implement a macro that automatically applies the inner-monomorphic-function pattern. It could be used like this:

```rust
// The attribute accepts the names of common traits that fit a many-input-types,
// one-output-type pattern (mainly Into and AsRef); it effectively moves the polymorphic
// conversion to the call site, so that the body doesn't need to be generic over the given traits.
#[monomorphic(Into)]
fn foo(x: impl Into<u32>, y: impl Into<u32>) {
    // function body will only be compiled once,
    // instead of for every combination of (typeof x) * (typeof y)
}
```

Because it's easy to implement (we already have `wrap_body`), easy to apply, and helps with monomorphization (which is one of the biggest causes of code bloat), this is probably one of the lowest-hanging fruit of compile-size optimizations. For now, I'll just put `wrap_body` in its own crate so that we'll have it handy when we get to the optimizing phase.

**Comment by kazcw** [comment]
@wdanilo 

> 2\. I've got an idea but I'm not sure yet what to think about it – what about creating a test that runs both `Vec::push` and our framework and if our framework runs longer than `x * time_of_vc_push` it fails? This way we will catch the situation when someone creates some changes to the code which break the perf. I believe that it would be possible to find an `x` which is big enough to not be affected by the time variance and small enough to catch bad modifications of this codebase in the future. What do you think about it? Do you think it's doable / worth it?

We could use [Criterion](https://bheisler.github.io/criterion.rs/book/index.html) as our microbenchmark runner--it supports regression testing, and adds smart statistical analysis and other features to `cargo bench`

**Comment by wdanilo** [comment]
> We could use [Criterion](https://bheisler.github.io/criterion.rs/book/index.html) as our microbenchmark runner--it supports regression testing, and adds smart statistical analysis and other features to `cargo bench`

Sure, but also we should merge it as soon as possible. Using Criterion with all these features sounds like a big task (but I might be wrong). Anyway, when can we merge this approx? :)

**Comment by kazcw** [comment]
@wdanilo Now! I'm not sure why it still says "Code owner review required" though--you are a code owner and have approved it, but I still cannot merge

**Comment by kazcw** [comment]
@mwu-tow Could you force-merge this? It says merge is blocked by "Code owner review" but I have an approval by a code owner; I think it is only blocked by CI. The CI failures are: some signature issue on Windows, definitely unrelated; and, no CHANGELOG entry, but this is internal tooling, so I don't think it needs one.

**Comment by vitvakatu** [comment]
@kazcw FYI, you can add `[ci no changelog needed]` anywhere in your PR description or in the last commit message and it will skip CHANGELOG.md check.

**Review by farmaazon — APPROVED** [review]
The merge is still blocked, because You've changed the `Cargo.toml` file, which is _not_ owned by @wdanilo.

The ownership works this way: to have merging PR enabled, for EACH file there should be an accept from AT LEAST ONE of its owners.

With my approval, it should work.

**Comment by kazcw** [comment]
@mwu-tow New build failures here (after I did a merge commit so CI would see the `[ci no changelog needed]` annotation)

**Comment by mwu-tow** [comment]
@kazcw The failures were due to Enso Standard Library example test accessing an external URL that was apparently not responsive at the time. Let's follow up with Engine Team on Discord.

**Pull request merged**
